### PR TITLE
[#7][#21] [iOS] [Integrate] As a user, I can see the log in screen

### PR DESCRIPTION
**Pull request by phongvhd93**
- Close:
  - #7 
  - #21 

## What happened 👀

- Add `LogInViewModel` and `LogInUseCase` to handle login flow.
- Add a String extension to validate the email. 
- Add `KMPNativeCoroutinesCombine` and `NativeCoroutine`  to wrap Kotlin Coroutine to Combine.
- Add a loading animation while calling API

## Insight 📝

When the user does not log in to the application, we will show a login screen. After a user logs in successfully, their credential data will be saved in KeyChain Access. 

## Proof Of Work 📹

https://user-images.githubusercontent.com/22606906/236715794-b18eb2a4-35d8-4b75-abed-00c33645fad1.mp4

### CODE REVIEW

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/LogIn/LogInCombineViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//  Copyright © 2023 Nimble. All rights reserved.
+//
+
+import Combine
+import Factory
+import Foundation
+import KMPNativeCoroutinesCombine
+import shared
+
+final class LogInCombineViewModel: ObservableObject {
```
Comment (markgravity) [inline iosApp/PhongKMMIC/Sources/Presentation/Modules/LogIn/LogInCombineViewModel.swift:15]: What do you think about using `LogInNativeViewModel` instead of `LogInCombineViewModel`?

File: iosApp/PhongKMMIC/Sources/Presentation/ViewModifiers/PrimaryTextFieldModifier.swift [hunk 2]
```diff
@@ -10,23 +10,28 @@ import SwiftUI
 
 struct PrimaryTextFieldModifier: ViewModifier {
 
+    @Binding var error: Bool
```
Comment (markgravity) [inline iosApp/PhongKMMIC/Sources/Presentation/ViewModifiers/PrimaryTextFieldModifier.swift]: What do you think about `hasError`?
The current name is quite ambiguous, I though we are passing an error message or object 🤔

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package co.nimblehq.avishek.phong.kmmic.data.remote.body
+
+import co.nimblehq.avishek.phong.kmmic.BuildKonfig
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+private const val GRANT_TYPE_PASSWORD = "password"
+
+@Serializable
+data class LogInRequestBody(
```
Comment (markgravity) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:10]: Please separate the Backend task to another PR next time, this PR is quite large and a bit hard to review now 🙏

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/LogIn/LogInCombineViewModel.swift (hunk 1 above)
Comment (phongvhd93) [reply iosApp/PhongKMMIC/Sources/Presentation/Modules/LogIn/LogInCombineViewModel.swift:15]: @markgravity, I'm following the pattern of this [example](https://github.com/rickclephas/KMP-NativeCoroutines), which also the library I'm using to wrap Kotlin Coroutine to Combine

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt (hunk 3 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:10]: I'll do that next time

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/di/extension/Start.kt [hunk 4]
```diff
@@ -9,3 +10,5 @@ fun KoinApplication.Companion.start(): KoinApplication = initKoin()
 
 val Koin.splashViewModel: SplashViewModel
     get() = get()
+val Koin.logInViewModel: LogInViewModel
+    get() = get()
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/di/extension/Start.kt:14]: Do we actually need these vals?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/usecase/LogInUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package co.nimblehq.avishek.phong.kmmic.domain.usecase
+
+import co.nimblehq.avishek.phong.kmmic.domain.model.Token
+import co.nimblehq.avishek.phong.kmmic.domain.repository.AuthenticationRepository
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.onEach
+
+interface LogInUseCase {
+
+    operator fun invoke(email: String, password: String): Flow<Token>
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/usecase/LogInUseCase.kt:9]: We can remove the blank line before the first line in interfaces. ⚠️ But keep it in the class.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/extension/String+Regex.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package co.nimblehq.avishek.phong.kmmic.helper.extension
+
+fun String.isValidEmail(): Boolean {
+
+    val expression = "\\b[\\w\\.-]+@[\\w\\.-]+\\.\\w{2,4}\\b"
+    val regex = Regex(expression)
+    return isNotEmpty() && matches(regex)
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/extension/String+Regex.kt]: We can use the `toRegex()` method here.
```suggestion
    return isNotEmpty() && matches(expression.toRegex())
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    private fun setLoadingState() {
+        mutableViewState.update {
+            LogInViewState(isLoading = true)
+        }
+    }
+
+    private fun handleLoginError(error: Throwable) {
+        mutableViewState.update {
+            LogInViewState(isLoading = false, error = error.toErrorMessage())
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:59]: I was wondering, what if we need to show custom error messages on the app side instead of the API-retuned or the default throwable messages?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    val viewState: StateFlow<LogInViewState> = mutableViewState
+
+    fun logIn(email: String, password: String) {
+        if (!validInput(email, password)) { return }
+        setLoadingState()
+        viewModelScope.launch {
+            logInUseCase(email = email, password = password)
+                .catch { error -> handleLoginError(error) }
+                .collect { loginSuccess() }
+        }
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt]: We can use the `launchIn(viewModelScope)` method to reduce nesting. Also, we can remove the curly braces around `return`.
```suggestion
if (!validInput(email, password)) return
        logInUseCase(email = email, password = password)
            .onStart { setLoadingState() }
            .catch { error -> handleLoginError(error) }
            .onEach { loginSuccess() }
            .launchIn(viewModelScope)
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    private lateinit var useCase: LogInUseCase
+
+    private val mockThrowable = Throwable("mock")
+    private val mockToken = Token(
+        "",
+        "",
+        0,
+        "",
+        0
+    )
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt]: I think this should work too making the setup simpler.
```suggestion
    private val mockToken = mock(classOf<Token>())
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            it.isSuccess shouldBe true
+            it.isLoading shouldBe false
+            it.isInvalidEmail shouldBe false
+            it.isInvalidPassword shouldBe false
+            it.error shouldBe null
+        }
+    }
+
+    @Test
+    fun `when logInUseCase return faile - it update view state with error message is mock`() = runTest {
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt]: How about this name?
```suggestion
    fun `when logInUseCase returns failure - it updates view state with error message accordingly`() = runTest {
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @AfterTest
+    fun tearDown() {
+        Dispatchers.resetMain()
+        mainThreadSurrogate.close()
+    }
+
+    @Test
+    fun `when logInUseCase return success - it update view state with isSuccess true`() = runTest {
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt]:
```suggestion
    fun `when logInUseCase returns success - it updates view state with isSuccess as true`() = runTest {
```

**Review by AVI5HEK — COMMENTED** [review]
A few concerns for now.

File: gradle.properties [hunk 12]
```diff
@@ -11,6 +11,12 @@ android.nonTransitiveRClass=true
 #MPP
 kotlin.mpp.enableCInteropCommonization=true
 kotlin.mpp.androidSourceSetLayoutVersion=2
+kotlin.mpp.enableGranularSourceSetsMetadata=true
+kotlin.native.enableDependencyPropagation=false
+
+kotlin.native.cacheKind=none
+kotlin.native.binary.memoryModel=experimental
+kotlin.native.binary.freezing=disabled
```
Comment (suho) [inline gradle.properties:19]: Why did you add these configurations?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt [hunk 13]
```diff
@@ -2,20 +2,24 @@ package co.nimblehq.avishek.phong.kmmic.data.remote.datasource
...
 
 class TokenRemoteDataSourceImpl(private val apiClient: ApiClient) : TokenRemoteDataSource {
 
     override fun refreshToken(body: RefreshTokenRequestBody): Flow<TokenApiModel> {
-        return apiClient.responseBody("/v1/oauth/token", HttpMethod.Post, body)
+        return apiClient.responseBody("api/v1/oauth/token", HttpMethod.Post, body)
+    }
+
+    override fun logIn(body: LogInRequestBody): Flow<TokenApiModel> {
+        return apiClient.responseBody("api/v1/oauth/token", HttpMethod.Post, body)
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: Please define the API path

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+import org.koin.core.component.inject
+
+object UITestLogout: KoinComponent {
+
+    private val localDataSource: TokenLocalDataSource by inject()
+
+    fun removeAccessToken() {
+        try {
+            localDataSource.removeToken()
+        } catch (e: Exception) {
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt]: Please fix this warning!

File: gradle.properties (hunk 12 above)
Comment (phongvhd93) [reply gradle.properties:19]: We're using kotlin 1.7.10 in this project, and those settings are required to keep Koin working properly.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt (hunk 13 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: @suho, did you mean a constant for the API path?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt (hunk 7 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:59]: Should we define a generic message in this case?  Do you have any suggestions?

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt (hunk 9 above)
Comment (phongvhd93) [reply shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt]: It doesn't work in this case. I got an error while trying to do that.

Screenshot 2023-05-08 at 14 38 19

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt (hunk 13 above)
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: @phongvhd93 Yes, currently, you are using the same endpoint string for 2 APIs

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/di/extension/Start.kt (hunk 4 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/di/extension/Start.kt:14]: Yes, I'm using it in this file [Container.swift](https://github.com/nimblehq/avishek-phong-kmm-ic/blob/e1b12fe6229bd55de3a1697c94e9441dcdc81f56/iosApp/PhongKMMIC/Sources/Supports/Extensions/DI/Container.swift).

File: iosApp/PhongKMMIC/Sources/Presentation/ViewModifiers/PrimaryTextFieldModifier.swift (hunk 2 above)
Comment (phongvhd93) [reply iosApp/PhongKMMIC/Sources/Presentation/ViewModifiers/PrimaryTextFieldModifier.swift]: solved in 27ddae6

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/usecase/LogInUseCase.kt (hunk 5 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/usecase/LogInUseCase.kt:9]: solved in 27ddae6

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt]: solved in 27ddae6

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt (hunk 8 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt]: solved in 27ddae6

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt (hunk 10 above)
Comment (phongvhd93) [reply shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt]: solved in 27ddae6

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt (hunk 11 above)
Comment (phongvhd93) [reply shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt]: solved in 27ddae6

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt (hunk 13 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: solved in 27ddae6

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
```
Comment (markgravity) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/LogInUseCaseTest.kt]: [Optional] We can make them in one line 🙏
```suggestion
    private val mockToken = Token("", "", 0, "", 0)
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+class LogInViewModelTest {
+
+    @Mock
+    val logInUseCase = mock(classOf<LogInUseCase>())
+
+    private lateinit var viewModel: LogInViewModel
+
+    private val mainThreadSurrogate = newSingleThreadContext("UI thread")
+    private val mockThrowable = Throwable("mock")
+    private val mockToken = Token(
```
Comment (markgravity) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/presentation.module/LogInViewModelTest.kt]: Same here

**Review by markgravity — APPROVED** [review]
LGTM with small suggestions

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt [hunk 17]
```diff
@@ -0,0 +1,6 @@
+package co.nimblehq.avishek.phong.kmmic.data.remote.constant
+
+object Endpoints {
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt]: The singular form `Endpoint` would be more consistent with the package names, enums, etc.
```suggestion
object Endpoint {
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt:1]: The idea I shared was to keep a file `Constant.kt` at the root and put various objects containing const vals in it. E.g.,
```kotlin
object Endpoint {
  // const val ...
}

object DateTimePattern {
  // const val ...
}

object SomethingElse {
  // const val ...
}
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt (hunk 7 above)
Comment (AVI5HEK) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:59]: @phongvhd93 the ideal way is to expose the `Throwable` to the view by changing `val error: String? = null` to `val error: Throwable? = null`. Then let the view decide which error message to show. We can put the error message in a custom `Throwable` and throw it from the use case. E.g.,

```kotlin
// Throw them from the use cases:
object InvalidEmail: Throwable(message = getPlatformString("invalid_email"))
object InvalidPassword: Throwable(message = getPlatformString("invalid_password"))

// Handle them in the view:
when(error) {
  is InvalidEmail -> {
    // update UI
  }
  is InvalidPassword -> {
    // update UI
  }
}
```
To do this,
1. We must move the logic of email and password validation to use cases. They will throw `InvalidEmail`, `InvalidPassword`, etc.
2. Get the localized strings from the platforms using expect-actual mechanism. [Here's](https://medium.com/xorum-io/simple-string-localization-in-kmm-beaffc5e7a17) an article regarding that.

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:59]: @AVI5HEK  I got your idea. However, this is quite big; can we bring it to another story? On iOS, I'm not using a key-value XML-based file to manage the localized string; we're using a 3rd party library and strong type for localized string. (e.g. `R.string.localizable.authenticationFieldsEmail()`)

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt:1]: solved in c4beaf7

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt (hunk 17 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/constant/Endpoints.kt]: solved in c4beaf7

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift [hunk 18]
```diff
@@ -6,13 +6,16 @@
 //  Copyright © 2023 Nimble. All rights reserved.
 //
 
+import Combine
 import Factory
+import KMPNativeCoroutinesCombine
+import KMPNativeCoroutinesCore
```
Comment (suho) [inline iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift]: Do we need these imports

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift [hunk 19]
```diff
@@ -6,13 +6,16 @@
...
 import shared
 import SwiftUI
 
 struct SplashView: View {
 
-    @InjectedObject(\.splashViewModel) private var viewModel
+    @InjectedObject(\.splashViewModel) private var viewModel: shared.SplashViewModel
```
Comment (suho) [inline iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift]: Do we need the prefix `shared.`?

File: iosApp/fastlane/Constants/Constant.swift [hunk 20]
```diff
@@ -18,7 +18,7 @@ enum Constant {
 
     static let stagingFirebaseAppId = "1:1005614911399:ios:8051ebc71d585a2abe3082"
     static let productionFirebaseAppId = "1:1005614911399:ios:0442661faf675297be3082"
-    static let firebaseTesterGroups = "" //"group_name, group_name"
+    static let firebaseTesterGroups = "" // "group_name, group_name"
```
Comment (suho) [inline iosApp/fastlane/Constants/Constant.swift]: Please revert this

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/Constant.kt [hunk 21]
```diff
@@ -0,0 +1,6 @@
+package co.nimblehq.avishek.phong.kmmic
+
+object Endpoint {
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/Constant.kt]: Please fix this comment

[the remark this answers is not in the document]
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/Constant.kt]: solved in f5b0877

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift (hunk 18 above)
Comment (phongvhd93) [reply iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift]: solved in f5b0877

File: iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift (hunk 19 above)
Comment (phongvhd93) [reply iosApp/PhongKMMIC/Sources/Presentation/Modules/Splash/SplashView.swift]: removed in f5b0877

File: iosApp/fastlane/Constants/Constant.swift (hunk 20 above)
Comment (phongvhd93) [reply iosApp/fastlane/Constants/Constant.swift]: solved in f5b0877

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt [hunk 22]
```diff
@@ -0,0 +1,19 @@
+package co.nimblehq.avishek.phong.kmmic.helper.ios
+
+import co.nimblehq.avishek.phong.kmmic.data.local.datasource.TokenLocalDataSource
+import io.ktor.util.logging.KtorSimpleLogger
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+object UITestLogout: KoinComponent {
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt:8]: What is this class for?

**Review by suho — APPROVED** [review]

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt (hunk 22 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/helper/ios/UITestLogout.kt:8]: It is for clearing the keychain access when running the UI test. If I don't clean up the keychain access, the test for the login screen won't pass.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    @SerialName("grant_type")
+    val grantType: String = GRANT_TYPE_PASSWORD,
+    @SerialName("email")
+    val email: String,
+    @SerialName("password")
+    val password: String,
+    @SerialName("client_id")
+    val clientId: String = BuildKonfig.CLIENT_ID,
+    @SerialName("client_secret")
+    val clientSecret: String = BuildKonfig.CLIENT_SECRET,
```
Comment (kaungkhantsoe) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:20]: Shouldn't we declare these as `nullable`?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    val isSuccess: Boolean = false,
+    val isLoading: Boolean = false,
+    val error: String? = null,
+    val isInvalidEmail: Boolean = false,
+    var isInvalidPassword: Boolean = false
+) {
+    constructor() : this(false)
+
+    constructor(error: String?) : this(false, false, error)
+}
```
Comment (kaungkhantsoe) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:18]: We don't need the constructors here.
```suggestion
)
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt (hunk 23 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:20]: I don't see any issue that needs to declare those properties as `nullable` 🤔. Did you experience any issues related to that before or benefit of it?

Comment (kaungkhantsoe) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:20]: @phongvhd93 I think we usually declare this as `nullable` so that the app won't crash when API made changes. Just a precaution. 😄

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt (hunk 24 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt:18]: @kaungkhantsoe, iOS module is using those constructor, please check the file `LogInCombineViewModel`

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt (hunk 23 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/LogInRequestBody.kt:20]: But it is not the object for API response 🤔 , it is the request body sent from the device.

**Review by AVI5HEK — APPROVED** [review]
